File: 2020/rust/day01/src/main.rs
/*
 * Day 01 - parts 1 and then modified for part 2
 * https://adventofcode.com/2020/day/1
*/

use std::env;
use std::fs::File;
use std::io;
use std::io::BufRead;
use std::path::Path;

fn main() {
    println!("Day 01!");

    let target_sum = 2020;

    let args: Vec<String> = env::args().collect();
    let input_filename = &args[1];

    let numbers = input_nums(input_filename).expect("Error getting numbers from input file");

    println!("");
    println!("---------------------------------------------------------");
    println!(":: Part 1");
    println!("---------------------------------------------------------");
    part1(&numbers[..], target_sum);

    println!("");
    println!("---------------------------------------------------------");
    println!(":: Part 2");
    println!("---------------------------------------------------------");
    part2(&numbers[..], target_sum);

    println!("")
}

// Return a vector of numbers from the input file
fn input_nums(filename: impl AsRef<Path>) -> io::Result<Vec<i32>> {
    let infile = File::open(filename).expect("Error opening file");
    let buff = io::BufReader::new(infile);
    Ok(buff
        .lines()
        .map(|l| l.expect("line error"))
        .map(|l| l.parse::<i32>().unwrap())
        .collect())
}

// Part 1 - Find the two numbers which produce the target sum,
// and then print the product of those two numbers
fn part1(numbers: &[i32], target_sum: i32) {
    let mut idx = 0;
    'outer: for i in numbers {
        for j in &numbers[idx..] {
            if i + j == target_sum {
                println!("Target sum found! {} + {} == {}", i, j, target_sum);
                println!("Product! {} * {} = {}", i, j, i * j);
                break 'outer;
            }
        }
        idx = idx + 1;
    }
}

// Part 2 - Find the three numbers which produce the target sum,
// and then print the product of those three numbers
fn part2(numbers: &[i32], target_sum: i32) {
    let mut idx = 0;
    'outer: for i in numbers {
        for j in &numbers[idx..] {
            for k in numbers {
                if i + j + k == target_sum {
                    println!("Target sum found! {} + {} + {} == {}", i, j, k, target_sum);
                    println!("Product! {} * {} * {} = {}", i, j, k, i * j * k);
                    break 'outer;
                }
            }
        }
        idx = idx + 1;
    }
}

File: 2020/golang/day01/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

// Return slice of expenses as ints
func get_expenses(input_file string) ([]int, error) {

	input, err := os.Open(input_file)
	if err != nil {
		return nil, err
	}
	defer input.Close()

	var expenses []int
	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		expense, err := strconv.Atoi(scanner.Text())
		if err != nil {
			// skip
		} else {
			expenses = append(expenses, expense)
		}
	}

	return expenses, nil

}

// Return the two expenses which sum to 2020
func part1(expenses []int) (int, int) {
	for _, x := range expenses {
		for _, y := range expenses {
			if x+y == 2020 {
				return x, y
			}
		}
	}
	return 0, 0
}

// Return the three expenses which sum to 2020
func part2(expenses []int) (int, int, int) {
	for _, x := range expenses {
		for _, y := range expenses {
			for _, z := range expenses {
				if x+y+z == 2020 {
					return x, y, z
				}
			}
		}
	}
	return 0, 0, 0
}

// Day01 program
func main() {

	expenses, err := get_expenses("./input")
	if err != nil {
		fmt.Println("error reading input")
		os.Exit(2)
	}

	a, b := part1(expenses)
	fmt.Printf("expenses: %v, %v; product: %v \n", a, b, a*b)

	a, b, c := part2(expenses)
	fmt.Printf("expenses: %v, %v, %v; product: %v \n", a, b, c, a*b*c)

}
